src/Features/CSharp/Portable/CodeRefactorings/SyncNamespace/CSharpSyncNamespaceCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,217 @@
+            {
+                return namespaceDecl.Members;
+            }
+            else if (node is CompilationUnitSyntax compilationUnit)
+            {
+                return compilationUnit.Members;
+            }
+            else
+            {
+                return default;
+            }
+        }
+
+        private static NameSyntax GetQualifiedNameSyntax(NameSyntax node) 
+            => node.Parent is QualifiedNameSyntax qualifiedName && qualifiedName.Right == node ? qualifiedName : node;
+
+        protected override bool TryGetReplacementSyntax(SyntaxNode reference, ImmutableArray<string> namespaceParts, out SyntaxNode old, out SyntaxNode @new)
+        {
+            if (reference is SimpleNameSyntax nameRef)
+            {
+                var outerMostNode = GetQualifiedNameSyntax(nameRef);
+                old = outerMostNode;
+
+                // If no namespace is specified, we just find and return the full NameSyntax.
+                if (outerMostNode == nameRef || namespaceParts.IsDefault)
+                {                        
+                    @new = outerMostNode;  
+                }
+                else
+                {
+                    var aliasQualifier = GetAliasQualifierOpt(outerMostNode);
+                    var qualifiedNamespaceName = CreateNameSyntax(namespaceParts, aliasQualifier, namespaceParts.Length - 1);
+
+                    // We might lose some trivia associated with nodes within `outerMostNode`.  
+                    @new = SyntaxFactory.QualifiedName(qualifiedNamespaceName, nameRef).WithTriviaFrom(outerMostNode);
+                } 
+                return true;
+            }
+            else
+            {
+                old = @new = null;
+                return false;
+            } 
+        }
+
+        protected override SyntaxNode CreateUsingDirective(ImmutableArray<string> namespaceParts)
+        {
+            var name = CreateNameSyntax(namespaceParts, aliasQualifier: null, namespaceParts.Length - 1);
+            return SyntaxFactory.UsingDirective(alias: null, name: name);
+        }
+
+        protected override string EscapeIdentifier(string identifier)
+            => identifier?.EscapeIdentifier();
+
+        private static string GetAliasQualifierOpt(NameSyntax name)
+        {                          
+            while (true)
+            {
+                switch (name.Kind())
+                {
+                    case SyntaxKind.QualifiedName:
+                        name = ((QualifiedNameSyntax)name).Left;
+                        continue;
+                    case SyntaxKind.AliasQualifiedName:
+                        return ((AliasQualifiedNameSyntax)name).Alias.Identifier.ValueText;
+                }
+
+                return null;
+            }
+        }
+        
+        private NameSyntax CreateNameSyntax(ImmutableArray<string> namespaceParts, string aliasQualifier, int index)
```

Comment:
these could all benefit from some explaantions as to why their necessary.

Most specifically, i'd like to understand if we've ever starting from an INAmespaceSymbol.  If so, i don't think any of this code is necessary.